File: Python/6-LinkedList/Algorithms/1-Floyd’sCycleDetection(Tortoise-and-Hare)/Floyd’sCycleDetection(Tortoise-and-Hare).py
class ListNode:
    """
    Represents a node in a singly linked list.

    Attributes:
        val: The value stored in the node.
        next: A reference to the next node in the list.
    """
    def __init__(self, val=0):
        self.val = val
        self.next = None

def detect_cycle_floyd(head):
    """
    Detects if a linked list has a cycle using Floyd's cycle-finding algorithm (tortoise and hare).

    Args:
        head: The head node of the linked list.

    Returns:
        True if a cycle is detected, False otherwise.

    Algorithm:
        - Uses two pointers, 'slow' and 'fast'.
        - 'slow' moves one step at a time, 'fast' moves two steps.
        - If there's a cycle, 'slow' and 'fast' will eventually meet.
    """
    slow, fast = head, head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True  # Cycle detected
    return False  # No cycle

def detect_cycle_set(head):
    """
    Detects if a linked list has a cycle using a set to store visited nodes.

    Args:
        head: The head node of the linked list.

    Returns:
        True if a cycle is detected, False otherwise.

    Algorithm:
        - Iterates through the list, adding each node to a set.
        - If a node is encountered that's already in the set, a cycle exists.
    """
    visited = set()
    while head:
        if head in visited:
            return True  # Cycle detected
        visited.add(head)
        head = head.next
    return False  # No cycle

def detect_cycle_dict(head):
    """
    Detects if a linked list has a cycle using a dictionary to store visited nodes.
    This method uses the id of the node as the key.

    Args:
        head: The head node of the linked list.

    Returns:
        True if a cycle is detected, False otherwise.

    Algorithm:
        - Iterates through the list, adding the id of each node to a dictionary.
        - If the id of a node is encountered that's already in the dictionary, a cycle exists.
    """
    visited = {}
    while head:
        if visited.get(id(head)):
            return True  # Cycle detected
        visited[id(head)] = True
        head = head.next
    return False  # No cycle

def detect_cycle_recursive(head, visited=None):
    """
    Detects if a linked list has a cycle using recursion and a set to store visited nodes.

    Args:
        head: The head node of the linked list.
        visited: A set to store visited nodes (used for recursion).

    Returns:
        True if a cycle is detected, False otherwise.

    Algorithm:
        - Uses recursion to traverse the list.
        - A set 'visited' keeps track of visited nodes.
        - If a node is encountered that's already in 'visited', a cycle exists.
    """
    if visited is None:
        visited = set()
    if head is None:
        return False  # End of list, no cycle
    if head in visited:
        return True  # Cycle detected
    visited.add(head)
    return detect_cycle_recursive(head.next, visited)

def detect_cycle_position_marker(head):
    """
    Detects if a linked list has a cycle by modifying the list itself.
    This method replaces each visited node's next pointer with a marker node.

    Args:
        head: The head node of the linked list.

    Returns:
        True if a cycle is detected, False otherwise.

    Algorithm:
        - Uses a marker node to track visited nodes.
        - For each node, if its next pointer points to the marker, a cycle exists.
        - Otherwise, the node's next pointer is set to the marker.
        -This approach modifies the original list.
    """
    marker = ListNode(-1)
    while head:
        if head.next == marker:
            return True  # Cycle detected
        next_node = head.next
        head.next = marker
        head = next_node
    return False  # No cycle

# Helper function to create a cycle for testing
def create_cycle(lst, pos):
    """
    Creates a linked list with a cycle for testing purposes.

    Args:
        lst: A list of values to create the linked list from.
        pos: The index of the node where the cycle starts (-1 for no cycle).

    Returns:
        The head node of the created linked list.
    """
    if not lst:
        return None
    head = ListNode(lst[0])
    nodes = [head]
    for val in lst[1:]:
        nodes.append(ListNode(val))
        nodes[-2].next = nodes[-1]
    if pos != -1:
        nodes[-1].next = nodes[pos]  # Create the cycle
    return head

# Test Cases
nodes = [3, 2, 0, -4]
head = create_cycle(nodes, 1)  # Create a cycle starting at the second node

print("Floyd’s Algorithm:", detect_cycle_floyd(head))
print("Set-based Approach:", detect_cycle_set(head))
print("Dictionary-based Approach:", detect_cycle_dict(head))
print("Recursive Approach:", detect_cycle_recursive(head))
print("Position Marker Approach:", detect_cycle_position_marker(head))

File: Java/6-LinkedList/Algorithms/1-Floyd’sCycleDetection(Tortoise-and-Hare)/Floyd’sCycleDetection(Tortoise-and-Hare).java
import java.util.HashSet;
import java.util.Set;



public class LinkedListCycleDetection {

    public static void main(String[] args) {
        // Create a sample linked list with a cycle for testing.
        ListNode head = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);
        ListNode node5 = new ListNode(5);

        head.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node5;
        node5.next = node2; // Create a cycle: 5 points back to 2.

        // Test the different cycle detection methods.
        System.out.println("Using Floyd's Tortoise and Hare (Method 1): Cycle detected: " + hasCycleFloyd(head));
        System.out.println("Using Floyd's Tortoise and Hare with Cycle Start (Method 2): Cycle detected: " + hasCycleFloydWithStart(head));

        System.out.println("Using HashSet (Method 3): Cycle detected: " + hasCycleHashSet(head));

        System.out.println("Using Visited Flag (Method 4): Cycle detected: " + hasCycleVisitedFlag(head));

        System.out.println("Using List Length Check (Method 5): Cycle detected: " + hasCycleLengthCheck(head));
    }

    // Method 1: Floyd's Tortoise and Hare Algorithm (Basic)
    //   - Uses two pointers, 'slow' and 'fast'.
    //   - 'slow' moves one node at a time, 'fast' moves two nodes at a time.
    //   - If there is a cycle, the two pointers will eventually meet.
    //   - Time Complexity: O(n), Space Complexity: O(1)
    public static boolean hasCycleFloyd(ListNode head) {
        if (head == null || head.next == null) {
            return false; // No cycle if the list is empty or has only one node.
        }

        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;      // Move slow pointer one step.
            fast = fast.next.next; // Move fast pointer two steps.
            if (slow == fast) {
                return true; // Cycle detected: slow and fast pointers meet.
            }
        }
        return false; // No cycle: fast pointer reached the end of the list.
    }

    // Method 2: Floyd's Tortoise and Hare Algorithm (Find Cycle Start)
    //   - Same as Method 1, but also finds the start node of the cycle.
    //   - After the pointers meet, move 'slow' back to the head.
    //   - Move both 'slow' and 'fast' one step at a time until they meet again.
    //   - The meeting point is the start of the cycle.
    public static boolean hasCycleFloydWithStart(ListNode head) {
        if (head == null || head.next == null) {
            return false;
        }

        ListNode slow = head;
        ListNode fast = head;
        ListNode meetingPoint = null; // Store the meeting point of slow and fast pointers

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                meetingPoint = slow; // Store the meeting point
                break;
            }
        }

        if (meetingPoint == null) {
            return false; // No cycle
        }

        // Find the start of the cycle
        slow = head;
        while (slow != meetingPoint) {
            slow = slow.next;
            meetingPoint = meetingPoint.next;
        }
        return true;
    }



    // Method 3: Using a HashSet
    //   - Stores each visited node in a HashSet.
    //   - If a node is encountered that is already in the HashSet, a cycle exists.
    //   - Time Complexity: O(n), Space Complexity: O(n)
    public static boolean hasCycleHashSet(ListNode head) {
        if (head == null) {
            return false;
        }

        Set<ListNode> visitedNodes = new HashSet<>();
        ListNode current = head;

        while (current != null) {
            if (visitedNodes.contains(current)) {
                return true; // Cycle detected: current node was already visited.
            }
            visitedNodes.add(current); // Add the current node to the HashSet.
            current = current.next;
        }
        return false; // No cycle: reached the end of the list.
    }

    // Method 4: Using a Visited Flag in the Node
    //   - Modifies the ListNode class (in-place modification).  //<- Important
    //   - Adds a 'visited' boolean flag to each node.
    //   - Set the flag to true when a node is visited.
    //   - If a visited node is encountered again, a cycle exists.
    //   - Time Complexity: O(n), Space Complexity: O(1) (excluding the space for the added flag)
    //   - Note: This method modifies the original linked list.  In some cases, you may not be allowed to modify the list.
    public static boolean hasCycleVisitedFlag(ListNode head) {
        if (head == null) {
            return false;
        }

        ListNode current = head;
        while (current != null) {
            if (current.val == Integer.MAX_VALUE) { // Using Integer.MAX_VALUE as a visited flag.  A boolean would be better, but avoids modifying the class.
                return true;
            }
            current.val = Integer.MAX_VALUE;  // Mark the current node as visited.
            current = current.next;
        }
        return false;
    }

    // Method 5: Cycle Detection by Length Check
    //  - Calculate the length of the linked list by traversing it.
    //  - If a cycle exists, the traversal might not terminate, or it will take a very long time.
    //  - We can set a maximum allowed length (e.g., 2n, where n is an estimated length).
    //  - If the traversal exceeds this limit, we assume there's a cycle.
    //  - This method is not very reliable, as the maximum length is arbitrary and depends on list
    //  - Time Complexity: O(n) in best case, O(infinite) in worst case (if no limit), Space Complexity: O(1)
    public static boolean hasCycleLengthCheck(ListNode head) {
        if (head == null) {
            return false;
        }

        int maxLength = 1000; // Arbitrary maximum length.  Should be related to expected list length.
        int length = 0;
        ListNode current = head;

        while (current != null && length <= maxLength) {
            length++;
            current = current.next;
        }

        return length > maxLength; // If length exceeds maxLength, assume a cycle.
    }
}

// Definition of a singly linked list node.
class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
        this.val = val;
        this.next = null;
    }
}
